src/dev/lupio/lupio_tmr.cc

```diff
@@ -0,0 +1,199 @@
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "dev/lupio/lupio_tmr.hh"
+
+#include "cpu/base.hh"
+#include "debug/LupioTMR.hh"
+#include "mem/packet_access.hh"
+#include "params/LupioTMR.hh"
+
+// Specific fields for CTRL
+#define LUPIO_TMR_IE    0x1
+#define LUPIO_TMR_PD    0x2
+
+// Specific fields for STAT
+#define LUPIO_TMR_EX    0x1
+
+namespace gem5
+{
+using namespace RiscvISA;
+LupioTMR::LupioTMR(const Params &params) :
+    BasicPioDevice(params, params.pio_size),
+    system(params.system),
+    tmrEvent([this]{ lupioTMRCallback(); }, name()) 
+{
+    start = curTick(); 
+    DPRINTF(LupioTMR, "LupioTMR initalized\n");
+}
+
+void LupioTMR::updateIRQ(int level)
+{
+    auto tc = system->threads[0];
+    // post an interrupt
+    if (level) {
+        tc->getCpuPtr()->postInterrupt(tc->threadId(),
+        ExceptionCode::INT_TIMER_MACHINE, 0);
+    }
+    // clear the interrupt
+    else {
+        tc->getCpuPtr()->clearInterrupt(tc->threadId(),
+        ExceptionCode::INT_TIMER_MACHINE, 0);
+
+    }
+}
+
+uint64_t
+LupioTMR::lupioTMRCurrentTime()
+{
+    return (curTick() - start) / sim_clock::as_int::ns; 
+}
+
+void
+LupioTMR::lupioTMRSet()
+{
+    startCycle = curCycle(); 
+    schedule(tmrEvent, Cycles(reload + curTick())); 
+}
+
+void
+LupioTMR::lupioTMRCallback()
+{
+    // Signal expiration 
+    expired = true;
+    if (ie) {
+        updateIRQ(1);
+    }
+
+    // If periodic timer, reload
+    if (pd && reload) {
+        lupioTMRSet();
+    }
+}
+
+uint64_t
+LupioTMR::lupioTMRRead(uint8_t addr, int size)
+{
+    uint64_t r = 0;
+    
+    switch (addr >> 2) {
+        case LUPIO_TMR_TIME:
+            r = lupioTMRCurrentTime();
+            DPRINTF(LupioTMR, "Read LUPIO_TMR_TME: %d\n", r);
+            break;
+        case LUPIO_TMR_LOAD:
+            r = reload;
+            DPRINTF(LupioTMR, "Read LUPIO_TMR_LOAD: %d\n", r);
+            break;
+        case LUPIO_TMR_STAT:
+            if (expired) {
+                r |= LUPIO_TMR_EX;
+            }
+
+            // Acknowledge expiration
+            expired = false;
+            DPRINTF(LupioTMR, "Read LUPIO_TMR_STAT: %d\n", r);
+            updateIRQ(0);
+            break;
+
+        default:
+            panic("Unexpected read to the LupioTMR device at address %#llx!",
+                    addr);
+            break;
+    }
+    return r;
+}
+
+void
+LupioTMR::lupioTMRWrite(uint8_t addr, uint64_t val64, int size)
+{
+    uint32_t val = val64;
+
+    switch (addr >> 2) {
+        case LUPIO_TMR_LOAD:
+                reload = val;
+                DPRINTF(LupioTMR, "Write LUPIO_TMR_LOAD: %d\n", reload);
+            break;
+
+        case LUPIO_TMR_CTRL:
+			ie = val & LUPIO_TMR_IE;
+			pd = val & LUPIO_TMR_PD;
```

Comment:
The indentation is not consistent for the lines above. Probably a mix of spaces and tabs.